rte/kernels-openacc/mo_rte_solver_kernels.F90

```diff
@@ -1197,5 +1198,310 @@ subroutine apply_BC_0(ncol, nlay, ngpt, top_at_1, flux_dn) bind (C, name="apply_
+    if(top_at_1) then
+      top_level = 1
+      lev_source_up => lev_source_dec
+      lev_source_dn => lev_source_inc
+    else
+      top_level = nlay+1
+      lev_source_up => lev_source_inc
+      lev_source_dn => lev_source_dec
+    end if
+
+    !$acc enter data copyin(d,tau,sfc_src,sfc_emis,lev_source_dec,lev_source_inc,lay_source,radn_dn)
+    !$acc enter data create(tau_loc,trans,source_dn,source_up,source_sfc,sfc_albedo,radn_up)
+    !$acc enter data attach(lev_source_up,lev_source_dn)
+
+    ! NOTE: This kernel produces small differences between GPU and CPU
+    ! implementations on Ascent with PGI, we assume due to floating point
+    ! differences in the exp() function. These differences are small in the
+    ! RFMIP test case (10^-6).
+    !$acc parallel loop collapse(3)
+    do igpt = 1, ngpt
+      do ilev = 1, nlay
+        do icol = 1, ncol
+          !
+          ! Optical path and transmission, used in source function and transport calculations
+          !
+          tau_loc(icol,ilev,igpt) = tau(icol,ilev,igpt)*D(icol,igpt)
+          trans  (icol,ilev,igpt) = exp(-tau_loc(icol,ilev,igpt))
+        end do
+      end do
+    end do
+
+    !$acc parallel loop collapse(2)
+    do igpt = 1, ngpt
+      do icol = 1, ncol
+      !
+      ! Transport is for intensity
+      !   convert flux at top of domain to intensity assuming azimuthal isotropy
+      !
+        radn_dn(icol,top_level,igpt) = radn_dn(icol,top_level,igpt)/(2._wp * pi * weight)
+      !
+        ! Surface albedo, surface source function
+      !
+        sfc_albedo(icol,igpt) = 1._wp - sfc_emis(icol,igpt)
+        source_sfc(icol,igpt) = sfc_emis(icol,igpt) * sfc_src(icol,igpt)
+      end do
+    end do
+
+
+    !
+    ! Source function for diffuse radiation
+    !
+    call lw_source_noscat(ncol, nlay, ngpt, &
+                          lay_source, lev_source_up, lev_source_dn, &
+                          tau_loc, trans, source_dn, source_up)
+
+    !
+    ! Transport
+    !
+    call lw_transport_1rescl(ncol, nlay, ngpt, top_at_1,  &
+                             tau_loc, scaling, trans, &
+                             sfc_albedo, source_dn, source_up, source_sfc, &
+                             radn_up, radn_dn)
+
+      ! Convert intensity to flux assuming azimuthal isotropy and quadrature weight
+      !
+      ! radn_dn(:,:,igpt) = 2._wp * pi * weight * radn_dn(:,:,igpt)
+      ! radn_up(:,:,igpt) = 2._wp * pi * weight * radn_up(:,:,igpt)
+  end subroutine lw_solver_1rescl
+  ! -------------------------------------------------------------------------------------------------
+  !
+  ! LW transport, no scattering, multi-angle quadrature
+  !   Users provide a set of weights and quadrature angles
+  !   Routine sums over single-angle solutions for each sets of angles/weights
+  !
+  ! ---------------------------------------------------------------
+ 
+  subroutine lw_solver_1rescl_GaussQuad(ncol, nlay, ngpt, top_at_1, nmus, Ds, weights, &
```

Comment:
It's an interesting question - do we need the Gaussian quadrature version of this code? I'm not sure we'd ever use more than one angle.

Reply:
2 quadrature angles provides more accurate result than a single angle. So I would rather keep the capability.